### PR TITLE
Update var_password_pam_remember_control_flag to allow multiple values in OL8

### PR DESCRIPTION
**Pull request by Xeicker**
#### Description:

- Update `var_password_pam_remember_control_flag`  to allow multiple values
- Update ansible, bash and OVAL to work with this new `var_password_pam_remember_control_flag`
- Update bash to use the `bash_ensure_pam_module_options` to cover all possible scenarios
- Update ansible regex to fix any control flag
- Update tests

#### Rationale:

- This makes the rule to accept multiple compliant scenarios
- The updates in fixes increase the scenarios which they can handle

### CODE REVIEW

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/correct_value_multiple_control_required.pass.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platforms = Oracle Linux 8
```
Comment (jan-cerny) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/correct_value_multiple_control_required.pass.sh:3]: this needs to be platform, not platforms
```suggestion
# platform = Oracle Linux 8
```

**Review by jan-cerny — DISMISSED** [review]

**Review by marcusburghardt — COMMENTED** [review]
I would suggest to consider the use of macros for the remediation.

Another point I am curious is about the acceptable controls. I am fine if this is the case for `ol8`, but I would like to remind that `requisite` and `required` are slightly different in this context. For example, in RHEL benchmarks, only one control is acceptable but the control was changed between RHEL7 and RHEL8. I wonder if the case is similar here.

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/argument_missing.fail.sh [hunk 2]
```diff
@@ -2,6 +2,8 @@
 # platform = Red Hat Enterprise Linux 7,Red Hat Virtualization 4,multi_platform_fedora,multi_platform_ol,multi_platform_wrlinux
 # packages = pam
 
+{{{ bash_package_remove("authselect") }}}
```
[the remark this answers is not in the document]
Comment (Xeicker) [reply linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/argument_missing.fail.sh:5]: This was thinking in an edge case of a system where authselect is available but not installed. Shouldn't we consider that?
In that case I should change the platform so it doesn't affect OL8

**Comment by Xeicker** [comment]
> Another point I am curious is about the acceptable controls. I am fine if this is the case for ol8, but I would like to remind that requisite and required are slightly different in this context. For example, in RHEL benchmarks, only one control is acceptable but the control was changed between RHEL7 and RHEL8. I wonder if the case is similar here.

Well from DISA STIG requirements `OL08-00-020221` & `OL08-00-020220`(and RHEL equivalents `RHEL-08-020221` &  `RHEL-08-020220`) it is not explicitly stated that any control must be there. And from man pages both `required` and `requisite` sounded reasonable. Could you explain with more detail why in RHEL8 only one of them is acceptable? may be the same applies to OL8

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/bash/shared.sh [hunk 3]
```diff
@@ -2,6 +2,8 @@
```
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/bash/shared.sh:2]: There is one PR (https://github.com/ComplianceAsCode/content/pull/9017) where the PAM macros for bash remediation were refactored to be more robust, simpler and flexible. And this rule was already updated to use the new PAM macros in that PR. Probably a rebase will be necessary if that PR get merged soon.

More context also in the https://github.com/ComplianceAsCode/content/issues/8959 issue.

**Comment by jan-cerny** [comment]
@marcusburghardt Could you please take a look at this? It's been rebased and as you recently modified the PAM and autheselect related things you have more context than me.

**Comment by Xeicker** [comment]
@marcusburghardt  I don't know why I thought the macros for ansible where also included in the PR you mentioned. 
Should I include the migration of ansible remediation to macros in this PR?

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/ansible/shared.yml [hunk 4]
```diff
@@ -125,8 +125,8 @@
     - name: Ensure the desired control value is updated in the custom profile
```
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/ansible/shared.yml:125]: @Xeicker , this meantime I refactored all macros related to PAM rules:
https://github.com/ComplianceAsCode/content/pull/9017
https://github.com/ComplianceAsCode/content/pull/9097

You can use them now. : )

**Comment by marcusburghardt** [comment]
> @marcusburghardt I don't know why I thought the macros for ansible where also included in the PR you mentioned. Should I include the migration of ansible remediation to macros in this PR?

Hi @Xeicker , yes. Please, use the new macros for the remediations. This way we can keep the same standard for all PAM related rules and make them easier to be maintained. The Ansible macros are very new (merged today). They would make our lives much easier, I hope. : )

**Comment by marcusburghardt** [comment]
> @marcusburghardt Could you please take a look at this? It's been rebased and as you recently modified the PAM and autheselect related things you have more context than me.

Sure @jan-cerny , I am assign this PR to myself. Thanks

**Comment by marcusburghardt** [comment]
@Xeicker , could you rabase the PR and solve the conflicts, please?

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/correct_value_multiple_control_required.pass.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platform = Oracle Linux 8
+# profiles = xccdf_org.ssgproject.content_profile_stig
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/correct_value_multiple_control_required.pass.sh:4]: For this test scenario, I would recommend to define the variable value to be testes instead of the profile where it is set. This way we make sure the test scenario is not impacted by any change in the profile. For example:
`variables = var_password_pam_remember=5,var_password_pam_remember_control_flag=ol8`

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/correct_value_multiple_control_requisite.pass.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platform = Oracle Linux 8
+# profiles = xccdf_org.ssgproject.content_profile_stig
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/correct_value_multiple_control_requisite.pass.sh:4]: For this test scenario, I would recommend to define the variable value to be testes instead of the profile where it is set. This way we make sure the test scenario is not impacted by any change in the profile. For example:
`variables = var_password_pam_remember=5,var_password_pam_remember_control_flag=ol8`

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/wrong_value_multiple_control_requisite.fail.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platform = Oracle Linux 8
+# profiles = xccdf_org.ssgproject.content_profile_stig
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/wrong_value_multiple_control_requisite.fail.sh:4]: For this test scenario, I would recommend to define the variable value to be testes instead of the profile where it is set. This way we make sure the test scenario is not impacted by any change in the profile. For example:
variables = var_password_pam_remember=5,var_password_pam_remember_control_flag=ol8

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/var_password_pam_remember_control_flag.var [hunk 8]
```diff
@@ -2,7 +2,10 @@ documentation_complete: true
 
 title: 'PAM pwhistory remember - control flag'
 
-description: 'Specify the control flag required for password remember requirement.'
+description: |-
+    'Specify the control flag required for password remember requirement. If multiple
+    values are allowed write them separated by commas as in "required,requisite",
+    for remediations the first value will be taken'
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/var_password_pam_remember_control_flag.var:8]: I had to read the description twice. Maybe we could subtly improve the reading. For example:
```
    'Specify the control flag required for password remember requirement. If multiple
    values are allowed, write them separated by commas as in "required,requisite".
    In case of multiple values, the first value will be taken for remediations.'
```

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/correct_value_multiple_control_required.pass.sh [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platform = Oracle Linux 8
+# profiles = xccdf_org.ssgproject.content_profile_stig
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/correct_value_multiple_control_required.pass.sh:4]: For this test scenario, I would recommend to define the variable value to be testes instead of the profile where it is set. This way we make sure the test scenario is not impacted by any change in the profile. For example:
`variables = var_password_pam_remember=5,var_password_pam_remember_control_flag=ol8`

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/correct_value_multiple_control_requisite.pass.sh [hunk 10]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platform = Oracle Linux 8
+# profiles = xccdf_org.ssgproject.content_profile_stig
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/correct_value_multiple_control_requisite.pass.sh:4]: For this test scenario, I would recommend to define the variable value to be testes instead of the profile where it is set. This way we make sure the test scenario is not impacted by any change in the profile. For example:
`variables = var_password_pam_remember=5,var_password_pam_remember_control_flag=ol8`

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/wrong_value_multiple_control_sufficient.fail.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# packages = pam
+# platform = Oracle Linux 8
+# profiles = xccdf_org.ssgproject.content_profile_stig
```
Comment (marcusburghardt) [inline linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_password_auth/tests/wrong_value_multiple_control_sufficient.fail.sh:4]: For this test scenario, I would recommend to define the variable value to be testes instead of the profile where it is set. This way we make sure the test scenario is not impacted by any change in the profile. For example:
`variables = var_password_pam_remember=5,var_password_pam_remember_control_flag=ol8`

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
I have some considerations regarding the test scenarios, to make them more independent from profiles. Also, the Ansible remediation is failing in some cases, but is due to an already fixed issue (https://github.com/ComplianceAsCode/content/issues/9141). If you rebase the PR, this should also be fixed. Thanks.

File: linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/correct_value_multiple_control_required.pass.sh (hunk 5 above)
Comment (Xeicker) [reply linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_pwhistory_remember_system_auth/tests/correct_value_multiple_control_required.pass.sh:4]: I wanted to do something similar, but it doesn't work in this way, if I set `var_password_pam_remember_control_flag=ol8` the value is literally "ol8" it doesn't take the value from the `.var` file

**Review by marcusburghardt — COMMENTED** [review]
I have successfully tested all the impacted rules. There are only some comments about wording where I think we can improve the experience for the readers. Also it would be good to rebase this PR so we can run all the tests considering the updates and fixes (https://github.com/ComplianceAsCode/content/issues/9141) on relevant macros. I wouldn't say these points are blockers to merge this PR, but I also think is a good moment to tackle them. @Xeicker , could you take a look and share your thoughts, please?

**Review by marcusburghardt — APPROVED** [review]

**Comment by marcusburghardt** [comment]
@jan-cerny , there are changes requested by you which seems to be already addressed in this PR. Could you update your review, please?
